chainindex/indexer.go

```diff
@@ -0,0 +1,450 @@
+	"github.com/filecoin-project/go-state-types/crypto"
+
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/types/ethtypes"
+	"github.com/filecoin-project/lotus/lib/sqlite"
+)
+
+var _ Indexer = (*SqliteIndexer)(nil)
+
+// IdToRobustAddrFunc is a function type that resolves an actor ID to a robust address
+type IdToRobustAddrFunc func(ctx context.Context, emitter abi.ActorID, ts *types.TipSet) (address.Address, bool)
+
+type SqliteIndexer struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	db *sql.DB
+	cs ChainStore
+
+	idToRobustAddrFunc IdToRobustAddrFunc
+
+	insertEthTxHashStmt                   *sql.Stmt
+	getNonRevertedMsgInfoStmt             *sql.Stmt
+	getMsgCidFromEthHashStmt              *sql.Stmt
+	insertTipsetMessageStmt               *sql.Stmt
+	updateTipsetToRevertedStmt            *sql.Stmt
+	hasTipsetStmt                         *sql.Stmt
+	updateTipsetToNonRevertedStmt         *sql.Stmt
+	removeTipsetsBeforeHeightStmt         *sql.Stmt
+	removeEthHashesOlderThanStmt          *sql.Stmt
+	updateTipsetsToRevertedFromHeightStmt *sql.Stmt
+	updateEventsToRevertedFromHeightStmt  *sql.Stmt
+	isTipsetMessageNonEmptyStmt           *sql.Stmt
+	getMinNonRevertedHeightStmt           *sql.Stmt
+	hasNonRevertedTipsetStmt              *sql.Stmt
+	updateEventsToRevertedStmt            *sql.Stmt
+	updateEventsToNonRevertedStmt         *sql.Stmt
+	getMsgIdForMsgCidAndTipsetStmt        *sql.Stmt
+	insertEventStmt                       *sql.Stmt
+	insertEventEntryStmt                  *sql.Stmt
+
+	gcRetentionEpochs   int64
+	reconcileEmptyIndex bool
+	maxReconcileTipsets int
+
+	mu           sync.Mutex
+	updateSubs   map[uint64]*updateSub
+	subIdCounter uint64
+
+	closeLk sync.RWMutex
+	closed  bool
+}
+
+func NewSqliteIndexer(path string, cs ChainStore, gcRetentionEpochs int64, reconcileEmptyIndex bool,
+	maxReconcileTipsets int) (si *SqliteIndexer, err error) {
+	db, _, err := sqlite.Open(path)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to setup message index db: %w", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	defer func() {
+		if err != nil {
+			_ = db.Close()
+			cancel()
+		}
+	}()
+
+	err = sqlite.InitDb(ctx, "chain index", db, ddls, []sqlite.MigrationFunc{})
+	if err != nil {
+		return nil, xerrors.Errorf("failed to init chain index db: %w", err)
+	}
+
+	si = &SqliteIndexer{
+		ctx:                 ctx,
+		cancel:              cancel,
+		db:                  db,
+		cs:                  cs,
+		updateSubs:          make(map[uint64]*updateSub),
+		subIdCounter:        0,
+		gcRetentionEpochs:   gcRetentionEpochs,
+		reconcileEmptyIndex: reconcileEmptyIndex,
+		maxReconcileTipsets: maxReconcileTipsets,
+	}
+	if err = si.prepareStatements(); err != nil {
+		return nil, xerrors.Errorf("failed to prepare statements: %w", err)
+	}
+
+	return si, nil
+}
+
+func (si *SqliteIndexer) Start() error {
+	si.wg.Add(1)
```

Comment:
what's going on with the waitgroup for the gc loop? maybe just a channel you can close instead since there's only one thing to watch?

Reply:
I usually like this pattern as it scales easily to spinning up more go-routines down the line. The only goal here is to ensure call go-routines spun up in the Indexer shut down before `Close` returns.